### ISSUE TITLE
Dimension issues Forge 2281 + MC 1.10.2 + build 86 + [OpenTerrainGenerator]

### ISSUE THREAD

**Issue by Beanius** [micdoodle8/Galacticraft#2701]
I'm pretty much getting a conflict with every dimension in my pack whenever i run it with Galacticraft build 86. First it was Futurepacks dimension ID: 100, i disabled the mod, then it caught onto random things dimension ID: -343800852.

It basically crashes on startup whenever your mods are enabled B)

**Comment by radfast** [comment]
I'm gonna need the full logs from you, obviously.

**Comment by Beanius** [comment]
Random things:
https://gist.github.com/Beanius/edcf1f4303db5c78bfd2883d56395e2e

Futurepack:
https://gist.github.com/Beanius/6350b52cc9de3a7e4a867e4061f4ea75

**Comment by radfast** [comment]
any chance you can get me **full logs** please, from game start up to this crash?

I'd like to see:
A)  Futurepack full log
B)  Full log of **exact same modlist as Futurepack** but **without** Galacticraft and MicdoodleCore

That should maybe show me what's different between the two situations.

Let's not worry about Random Things for now, that must be the same issue - only the number is different.

**Comment by Beanius** [comment]
Ait on it! Had to work irl, gonna launch it twice, and add in the full logs B)

**Comment by Beanius** [comment]
With futurepack
https://gist.github.com/Beanius/f2bf4f0d2295cbd829f21f1e074bfe95
-------------------------------------------------------------------------------
without Galacticraft and MicdoodleCore
https://gist.github.com/Beanius/70a974531fc72e5f93786318801d761f

**Comment by radfast** [comment]
Thanks for the full logs.

Looking at the logs, it's likely that a third mod - not Galacticraft nor FuturePack - is messing this up.

Please can you try without **Just Enough Dimensions** and (if that doesn't fix it) also try without **rftoolsdim**.

sorry about all the restarts - it's a problem when you have conflicts with such a large modlist

**Comment by Beanius** [comment]
Do you want me to disable Futurepack on the launches without JED and RFTD?

**Comment by Beanius** [comment]
with futurepack minus Just enough dimensions and RFtoolsdim
https://gist.github.com/Beanius/538978f97487d1ddcbb0bc671b8f4a3a
----------------------------------------------------------------------------------
it crashed with the dimension: 100 again. Had Futurepack on hmm

**Comment by Beanius** [comment]
Ait. I made a test pack with 9 mods.. BetterFps. Futurepack. Galacticraftcore, planets, micdoodleCore, optifine, xaero's minimap + Open Terrain Generator and Biome bundle.

It crashes if i launch the pack with OTG/BD + Futurepack together. But if i disable one of them it succeeds.

**Comment by Beanius** [comment]
Succeded in launch with big modlist minus OTG/Biome Bundle.
https://gist.github.com/Beanius/72f767d3fbc8bf953964eb345a3aaf1c
--------------------------------------------------------------------------------

**Comment by radfast** [comment]
good job!

That's these two?
>
     OpenTerrainGenerator-1.10.2-v17.jar
     Biome_Bundle-v5.jar

I can test their compatibility with us, now that you've pointed me to the cause :)

Although Galacticraft is currently labelled "BETA" I'd be surprised if our dimension handling code is at fault....  We've been transferring players between dimensions for 4 years now, and most of the issues have been sorted out - even on Bukkit / Thermos / Sponge / Bungee servers.

**Comment by Beanius** [comment]
Indeed! it's the favorable terrain mods i use in my packs B) Hope they will be compatible with Galacticraft once it's stable to hit the 'release' stage c:

**Comment by radfast** [comment]
Like I said, fault may be in those mods, not in Galacticraft.  I am investigating further ...

**Comment by radfast** [comment]
The bug is in OpenTerrainGenerator, these lines in their code class ``` com.khorn.terraincontrol.forge.TCPlugin```:
>

    private void registerDimensions()
    {
        for (int i = 2; i < 100; i++) {
            if (!DimensionManager.isDimensionRegistered(i)) {
                DimensionManager.registerDimension(i, DimensionType.register("OTGDim", "OTG", i,  WorldProviderTC.class, false));
            }
        }
    }

They need to include a ```try ... catch``` statement in there, around the line of code beginning ```DimensionManager.registerDimemsion(```.

Without this ```try ... catch``` then a combination of the three  technical things explained below, working together, causes occasional, random errors.

**In layman's terms, you're seeing the error because OTG tries to register 98 similar dimensions very very quickly, and sometimes the code just can't keep up with itself...**

Technical explanation:

- DimensionType.register() uses Forge's ```EnumHelper.addEnum()``` method
- [EnumHelper.addEnum()](https://github.com/MinecraftForge/MinecraftForge/blob/1.11.x/src/main/java/net/minecraftforge/common/util/EnumHelper.java#L261-L391) is non-standard dangerous code using reflection to directly manipulate private fields inside a Java ```Enum``` class
- The JVM (Java Virtual Machine) makes runtime optimisations, these shift into higher gear after the code has repeatedly gone over the same place again and again and again.  Looks like OTG registering 98 dimension types in a loop is enough to make that happen!  The JVM optimiser then somehow optimises out some of the changes made by EnumHelper, I guess EnumHelper is just too smart for it.  You can tell it's the JVM optimiser because the Enum is seen to have different sizes when referenced in different methods, see code snippet below.

Code snippet which simulates the erroring part of OTG's code, and proves that it's the JVM.  Running ```breakDimensions()``` in this code snippet in any mod's pre-init code will probably error (tested in a 1.10.2 development environment) - if coders testing this can't make it error, just try continuing the outer for loop for longer, like until i == 300 or more. 

    private void breakDimensions()
    {
        for (int i = 2; i < 100; i++)
        {
            if (!DimensionManager.isDimensionRegistered(i))
            {
                DimensionType dt = DimensionType.register("OTGDim" + i, "OTG", i, WorldProviderEnd.class, false);
                getById(dt.getId(), DimensionType.values().length);
            }
        }
    }

    public static DimensionType getById(int id, int compareSize)
    {
        int valuesSize = DimensionType.values().length;
        if (valuesSize != compareSize) throw new IllegalArgumentException("Wow, this should be impossible! Enum values size mismatch : " + compareSize + " vs. " + valuesSize);

        for (DimensionType dimensiontype : DimensionType.values())
        {
            if (dimensiontype.getId() == id)
            {
                return dimensiontype;
            }
        }

        return null;
    }

**None of the above has any connection with Galacticraft.**

I can't explain why OTG only errors in the way described when you have Galacticraft in your modpack and not otherwise, but we are **definitely** not doing anything weird to trigger this.  Could be that Galacticraft somehow works the JVM optimiser harder than other mods, so that the JVM is primed to optimise the OTG code into erroring, in the way explained above...

Whatever, this is **definitely** not a Galacticraft bug.

Looks like after OTG has done this stuff, then any other mod trying to register dimensions (FuturePack, Random Things) will encounter similar errors.

Possible fixes:

- [best] ask the OTG developers to find a different way, instead of registering 98 dimension types all at once - you're welcome to show them this discussion
- report this to Forge and see if Forge are interested to make their ```EnumHelper``` less dangerous
- report this as a bug in the Java 8 optimiser - though I'm not sure what their attitude will be to Forge's ```EnumHelper```, which is very non-standard code!
- patch OTG yourself
- remove OTG from your modpack

**Comment by radfast** [comment]
I've been able to reproduce this OTG problem in the simplest of example mods, so I have [reported this to Forge](https://github.com/MinecraftForge/MinecraftForge/issues/3885) - let's see what they say.

**Comment by Beanius** [comment]
Ait! appreciate it a lot ^^. Hopefully they get to fix it asap :3

**Comment by Beanius** [comment]
oh, and i checked the issue over at forge. It was weird cause you found that i was on java 8 25 right, but my computer has been on 131 since April 27 hmm

**Comment by radfast** [comment]
Reported also to OpenTerrainGenerator devs: https://github.com/PG85/OpenTerrainGenerator/issues/22

**Comment by Beanius** [comment]
Hopefully they'll figure out the problem 👍

**Comment by Beanius** [comment]
Thanks a lot with sending the issues ahead to OTG and Forge! You have been of the greatest help, Truly appreciate it !! c:
